playground/src/commonMain/kotlin/Metadata.kt

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2015-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import io.ktor.utils.io.core.*
+import io.rsocket.kotlin.*
+import io.rsocket.kotlin.core.*
+import io.rsocket.kotlin.metadata.*
+import io.rsocket.kotlin.payload.*
+
+@OptIn(ExperimentalMetadataApi::class)
+public fun metadata() {
+
+    //writing
+
+    val routing = RoutingMetadata("tag1", "route1", "something")
+    val tracing = ZipkinTracingMetadata(
+        kind = ZipkinTracingMetadata.Kind.Sample,
+        traceId = 123L,
+        spanId = 123L
+    )
+    val cm1 = buildCompositeMetadata {
+        add(routing)
+        add(tracing)
+    }
+    // or
+    val cm2 = CompositeMetadata(routing, tracing)
+
+    //all lambdas are inline, and don't create another objects for builders
+    //so no overhead comparing to using plain Payload(data, metadata) function
+    val payload = buildPayload {
+        data("Some data")
+        metadata(cm2)
+        // or
+        metadata(cm1)
+        // or
+        compositeMetadata {
+            add(routing)
+            add(tracing)
+
+            add(WellKnownMimeType.ApplicationAvro, buildPacket { writeText("some custom metadata building") })
+            //or
+            val raw = RawMetadata(WellKnownMimeType.ApplicationAvro, ByteReadPacket.Empty)
+
+            add(raw)
+        }
+    }
+
+    //reading
+    //property types are not needed, just for better online readability
+    val cm: CompositeMetadata = payload.metadata?.read(CompositeMetadata) ?: return
+
+    val rmr = RawMetadata.reader(WellKnownMimeType.ApplicationAvro)
+
+
+    val rm: RoutingMetadata = cm.get(RoutingMetadata)
+    val tm: ZipkinTracingMetadata = cm.get(ZipkinTracingMetadata)
+
+    //or
+    val rm1: RoutingMetadata = cm[RoutingMetadata]
```

Comment:
Think examples like this are better through well known extension methods

Reply:
Can you explain a little?
Or, show better example?

Reply:
cm.route()

Reply:
see above.

Reply:
so, you propose to add something like `val Payload.routingMetadata: RoutingMetadata` or `fun Payload.routingMetadata(): RoutingMetadata` which can be read once, right?
and same accessors for other out-of-the-box metadatas?

Reply:
Yes

Reply:
Maybe it's weird on Payload, but as an extension method you choose to import it.

Also the domain specific helper could hide any logic to check is CompositeMetadata vs single (non-composite) types, or potentially include defaults that make sense for that type.

Reply:
looks ok, will add it

Reply:
Let's not overthink it. Let's land and iterate for now. I use the routing rules from client side, so let's see that in the context of rsocket-cli and my own project.

Reply:
Ok, So may be, let's merge it now and you will check it
If so, we can at least release 0.11.0 with those changes already done
and continues with incremental releases per feature